server/src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -265,106 +260,126 @@ public ContentAccessCertificate getCertificate(Consumer consumer)
             return null;
         }
 
-        Environment env = this.environmentCurator.getConsumerEnvironment(consumer);
-
         ContentAccessCertificate existing = consumer.getContentAccessCert();
-        OwnerEnvContentAccess oeca = this.ownerEnvContentAccessCurator
-            .getContentAccess(owner.getId(), env == null ? null : env.getId());
 
-        StringBuilder pem = new StringBuilder();
-        boolean regenerate = false;
-
-        // If the cert was last updated before the org's last content update, or the cert has expired,
-        // or the payload has changed, regenerate the header.
-        // TODO: we should probably just store this all in one thing rather than two things.
-        if (existing != null) {
-            Date now = new Date();
-            Date expiration = existing.getSerial().getExpiration();
-            Date contentUpdate = owner.getLastContentUpdate();
+        ContentAccessCertificate result;
+        if (existing == null) {
+            result = createNewScaCertificate(consumer, owner);
+        }
+        else {
+            result = updateScaCertificate(consumer, owner, existing);
+        }
 
-            if (expiration.before(now) || !contentUpdate.before(existing.getUpdated()) ||
-                oeca == null || !contentUpdate.before(oeca.getUpdated())) {
+        return result;
+    }
 
-                consumer.setContentAccessCert(null);
-                this.contentAccessCertificateCurator.delete(existing);
+    private ContentAccessCertificate updateScaCertificate(Consumer consumer, Owner owner,
+        ContentAccessCertificate existing) throws GeneralSecurityException, IOException {
+        Date now = new Date();
+        Date expiration = existing.getSerial().getExpiration();
 
-                regenerate = true;
-            }
+        // regenerate x509
+        if (expiration.before(now)) {
+            Validity oneYearValidity = Validity.oneYear();
+            KeyPair keyPair = this.keyPairCurator.getConsumerKeyPair(consumer);
+            CertificateSerial serial = createSerial(oneYearValidity);
+            existing.setSerial(serial);
+            existing.setCert(createX509Cert(consumer, owner, serial, keyPair, oneYearValidity));
+            this.contentAccessCertificateCurator.saveOrUpdate(existing);
+        }
+        // regenerate content
+        Date contentUpdate = owner.getLastContentUpdate();
+        if (!contentUpdate.before(existing.getUpdated())) {
+            existing.setContent(this.createPayloadAndSignature(owner, consumer));
+            this.contentAccessCertificateCurator.saveOrUpdate(existing);
         }
+        return existing;
+    }
 
-        if (existing == null || regenerate) {
-            log.info("Generating new SCA certificate for organization \"{}\"...", owner.getKey());
+    private ContentAccessCertificate createNewScaCertificate(Consumer consumer, Owner owner)
```

Comment:
Given that we're using merge, saveOrUpdate, and other such state-agnostic persistence methods, is there any reason this method needs to exist individually? It seems separating the create and update operations isn't doing much for us, and the two could easily just be a single create-or-update style method as it was before.

There's a ton of abstraction happening here and lots of jumping around the code for no real gain.

Reply:
I shuffled a bit more with the methods and inlined some of them. I think that splitting it helps a bit with navigation. There is a lot of low level thing happening with cert generation. This way you can get an overview of what the getCertificate method does and move to details of creation/update if you need to.